tensorflow/compiler/xla/service/gpu/gpu_layout_assignment_test.cc

```diff
@@ -347,6 +355,31 @@ ENTRY entry {
   expect_layout(call_0->operand(1)->shape(), {1, 2, 0});
 }
 
+TEST_F(LayoutAssignmentTest, ConvCuDNNBF16) {
+  if (!GetCudaComputeCapability().IsAtLeast(se::CudaComputeCapability::VOLTA)) {
+    GTEST_SKIP() << "Conv with Bfloat16 uses NHWC layout for "
+                    "architectures with Tensor Cores.";
+  }
+
+  const char* hlo = R"(
+
+  HloModule jit_conv_general_dilated
+
+  ENTRY main.4 {
+    Arg_0.1 = bf16[1,64,64,16]{3,2,1,0} parameter(0), sharding={replicated}
+    Arg_1.2 = bf16[3,3,16,32]{3,2,1,0} parameter(1), sharding={replicated}
+    ROOT convolution.3 = bf16[1,64,64,32]{3,2,1,0} convolution(Arg_0.1, Arg_1.2), window={size=3x3 pad=1_1x1_1}, dim_labels=b01f_01io->b01f, metadata={op_name="jit(conv_general_dilated)/jit(main)/conv_general_dilated[window_strides=(1, 1) padding=((1, 1), (1, 1)) lhs_dilation=(1, 1) rhs_dilation=(1, 1) dimension_numbers=ConvDimensionNumbers(lhs_spec=(0, 3, 1, 2), rhs_spec=(3, 2, 0, 1), out_spec=(0, 3, 1, 2)) feature_group_count=1 batch_group_count=1 lhs_shape=(1, 64, 64, 16) rhs_shape=(3, 3, 16, 32) precision=None preferred_element_type=None]" source_file="/usr/local/lib/python3.8/dist-packages/flax/linen/linear.py" source_line=438}
+  }
+)";
+
+  MatchOptimizedHlo(hlo, R"(
```

Comment:
This doesn't seem to work. The Filecheck input we get is:

```
HloModule jit_conv_general_dilated, entry_computation_layout={(bf16[1,64,64,16]{3,2,1,0},bf16[3,3,16,32]{3,2,1,0})->bf16[1,64,64,32]{3,2,1,0}}

%fused_computation (param_0.3: f32[1,32,64,64]) -> bf16[1,64,64,32] {
  %param_0.3 = f32[1,32,64,64]{3,2,1,0} parameter(0)
  %bitcast.17 = f32[32,64,64]{2,1,0} bitcast(%param_0.3)
  %convert.8 = bf16[32,64,64]{2,1,0} convert(%bitcast.17)
  %transpose.3 = bf16[64,64,32]{2,1,0} transpose(%convert.8), dimensions={1,2,0}
  ROOT %bitcast.16 = bf16[1,64,64,32]{3,2,1,0} bitcast(%transpose.3)
}

%fused_computation.1 (param_0.5: bf16[3,3,16,32]) -> f32[32,16,3,3] {
  %param_0.5 = bf16[3,3,16,32]{3,2,1,0} parameter(0)
  %convert.9 = f32[3,3,16,32]{3,2,1,0} convert(%param_0.5)
  ROOT %transpose.4 = f32[32,16,3,3]{3,2,1,0} transpose(%convert.9), dimensions={3,2,0,1}
}

%fused_computation.2 (param_0.9: bf16[1,64,64,16]) -> f32[1,16,64,64] {
  %param_0.9 = bf16[1,64,64,16]{3,2,1,0} parameter(0)
  %bitcast.19 = bf16[64,64,16]{2,1,0} bitcast(%param_0.9)
  %convert.10 = f32[64,64,16]{2,1,0} convert(%bitcast.19)
  %transpose.5 = f32[16,64,64]{2,1,0} transpose(%convert.10), dimensions={2,0,1}
  ROOT %bitcast.18 = f32[1,16,64,64]{3,2,1,0} bitcast(%transpose.5)
}

ENTRY %main.4 (Arg_0.1: bf16[1,64,64,16], Arg_1.2: bf16[3,3,16,32]) -> bf16[1,64,64,32] {
  %Arg_0.1 = bf16[1,64,64,16]{3,2,1,0} parameter(0), sharding={replicated}
  %fusion.2 = f32[1,16,64,64]{3,2,1,0} fusion(%Arg_0.1), kind=kInput, calls=%fused_computation.2
  %Arg_1.2 = bf16[3,3,16,32]{3,2,1,0} parameter(1), sharding={replicated}
  %fusion.1 = f32[32,16,3,3]{3,2,1,0} fusion(%Arg_1.2), kind=kLoop, calls=%fused_computation.1
  %cudnn-conv.1 = (f32[1,32,64,64]{3,2,1,0}, u8[51328]{0}) custom-call(%fusion.2, %fusion.1), window={size=3x3 pad=1_1x1_1}, dim_labels=bf01_oi01->bf01, custom_call_target="__cudnn$convForward", metadata={op_name="jit(conv_general_dilated)/jit(main)/conv_general_dilated[window_strides=(1, 1) padding=((1, 1), (1, 1)) lhs_dilation=(1, 1) rhs_dilation=(1, 1) dimension_numbers=ConvDimensionNumbers(lhs_spec=(0, 3, 1, 2), rhs_spec=(3, 2, 0, 1), out_spec=(0, 3, 1, 2)) feature_group_count=1 batch_group_count=1 lhs_shape=(1, 64, 64, 16) rhs_shape=(3, 3, 16, 32) precision=None preferred_element_type=None]" source_file="/usr/local/lib/python3.8/dist-packages/flax/linen/linear.py" source_line=438}, backend_config="{\"algorithm\":{\"algo_id\":\"12\",\"math_type\":\"DEFAULT_MATH\",\"tuning_knobs\":{\"5\":\"1\",\"6\":\"0\",\"7\":\"1\",\"10\":\"1\"},\"is_cudnn_frontend\":true,\"workspace_size\":\"51328\"},\"conv_result_scale\":1,\"activation_mode\":\"0\",\"side_input_scale\":0}"
  %get-tuple-element.1 = f32[1,32,64,64]{3,2,1,0} get-tuple-element(%cudnn-conv.1), index=0
  ROOT %fusion = bf16[1,64,64,32]{3,2,1,0} fusion(%get-tuple-element.1), kind=kInput, calls=%fused_computation
}
```

Reply:
Ok, I figured it out, this test requires actually AMPERE to work, because BF16 support for convolutions is tied to that. It was running on a VOLTA machine and then it didn't work. I will fix this and then get this PR merged.

Reply:
Let me run some testing on Volta on my side as well. I was expecting this test to pass on Volta.

Reply:
On Volta, the convolution gets converted to f32, because it seems bf16 is only supported on Ampere or later.

Reply:
Thanks, I realized that, I would hence need to avoid defaulting to NHWC for Bfloat16 on Volta , and leave the layout as NCHW. I would need to push a new PR for changing this?

Reply:
When layout assignment is run, the conversion to f32 would have happened already. So I don't think any extra work is needed.